lang/udf_cache.hh

```diff
@@ -0,0 +1,208 @@
+/*
+ * Copyright (C) 2022-present ScyllaDB
+ */
+
+/*
+ * SPDX-License-Identifier: AGPL-3.0-or-later
+ */
+
+#pragma once
+
+#include "cql3/functions/function_name.hh"
+#include "utils/overloaded_functor.hh"
+#include "wasmtime.hh"
+#include "lang/wasm.hh"
+#include <seastar/core/shared_ptr.hh>
+#include <seastar/core/scheduling.hh>
+#include "utils/loading_cache.hh"
+#include "utils/hash.hh"
+#include "utils/overloaded_functor.hh"
+#include <chrono>
+#include <cstdint>
+#include <queue>
+#include <unordered_map>
+
+extern logging::logger wasm_logger;
+
+namespace wasm {
+
+struct udf_instance {
+    wasmtime::Store store;
+    wasmtime::Instance instance;
+    wasmtime::Func func;
+    wasmtime::Memory memory;
+};
+
+// For each UDF full name and a scheduling group, we store a future that is ready
+// when no other task uses the same UDF instance. When an instance is retrieved
+// from the cache, the future is replaced with a new one that becomes ready when
+// the instance is returned. To avoid conflicts between tasks, each time an 
+// instance is requested, the future associated with it is removed from the cache
+// and a new one is added. When the current instance user is done with it, the
+// future that was associated with it is becomes ready.
+// This way, an instance stays in the cache as long as it isn't used by any task
+// and if there are tasks that want to use it, they get it immediately after
+// the previous user is finished.
+class udf_cache {
+public:
+    struct stats {
+        uint64_t cache_hits = 0;
+        uint64_t cache_misses = 0;
+        uint64_t cache_blocks = 0;
+    };
+
+    static stats& shard_stats() {
+        static thread_local stats _stats;
+        return _stats;
+    }
+
+private:
+    using cache_key_type = std::pair<seastar::scheduling_group, cql3::functions::function_name>;
+    struct cache_entry_type {
+        mutable udf_instance inst;
+    };
+    struct udf_cache_entry_size {
+        size_t operator()(cache_entry_type& val) {
+            return val.inst.memory.size(val.inst.store);
```

Comment:
What about the bytecode and generated code size? Can we include them too?

Reply:
I didn't find any other way of checking the amount of memory used by an instance in wasmtime. 
From what I did find, in an (module) Instance there are instantiations of all module components, most importantly Functions but also Tables and Memories etc. The size of an Instance depends on all of them, some can be also lazily evaluated. That's probably why there isn't a general `size` method
There is something called `InstanceLimits` that allows limiting the space allocated for an instance, but it has to be used in combination with a `PoolingAllocator` which is a kind of an allocator that reserves resources in advance and reuses the pool for different instances (we're not using it currently but it has an interesting default value for total instance size - 1MB)

Reply:
We'll need to address it, otherwise we can have an accidental denial of service attack by too-large wasm programs. Anyway we can refine it later. To start with, bias the memory calculations with a constant value, say 100k (or something that fits your experience) so we don't allow an infinite amount of instances if they use tiny amounts of memory.

Reply:
We're counting the memory in wasm pages (64k) and a simple UDF that was using wasi with std collections took quite a bit less than that, so used 1 page for the bias